airmail_indexer/src/osm_osmx.rs

```diff
@@ -0,0 +1,164 @@
+use std::collections::HashMap;
+
+use airmail::poi::ToIndexPoi;
+use airmail_indexer::error::IndexerError;
+use anyhow::Result;
+use crossbeam::channel::Sender;
+use log::{debug, info, warn};
+use osmx::{Database, Locations, Transaction, Way};
+
+use crate::osm::OsmPoi;
+
+pub struct OSMExpressLoader<'db> {
+    sender: Sender<ToIndexPoi>,
+    transaction: Transaction<'db>,
+}
+
+impl<'db> OSMExpressLoader<'db> {
+    pub fn new(db: &'db Database, sender: Sender<ToIndexPoi>) -> Result<Self> {
+        // Share the transaction within the loader
+        let transaction = Transaction::begin(db).map_err(IndexerError::from)?;
+
+        Ok(Self {
+            sender,
+            transaction,
+        })
+    }
+
+    fn locations(&self) -> Result<Locations> {
+        let locations = self.transaction.locations().map_err(IndexerError::from)?;
+        Ok(locations)
+    }
+
+    // /// Option 1 - Get the centroid of the way
+    // ///
+    // /// This is slow as it requires all nodes to be fetched, then all locations to be
+    // /// fetched. This is a lot of seeks and reads.
+    // fn way_centroid(&self, way: &Way) -> Option<(f64, f64)> {
+    //     let locations = self.locations().ok()?;
+
+    //     // Fetch all nodes, driving the iterator to completion at once
+    //     let positions = way.nodes().collect::<Vec<_>>();
+
+    //     // Lookup each position
+    //     let node_positions: Vec<Coord> = positions
+    //         .iter()
+    //         .filter_map(|node| {
+    //             let node = locations.get(*node)?;
+    //             Some(Coord::from((node.lon(), node.lat())))
+    //         })
+    //         .collect();
+
+    //     if node_positions.is_empty() {
+    //         debug!("Empty node_positions");
+    //     }
+    //     let linestring = LineString::new(node_positions);
+    //     let polygon = Polygon::new(linestring, vec![]);
+    //     let centroid = polygon.centroid();
+    //     if centroid.is_none() {
+    //         debug!("No centroid for way");
+    //     }
+    //     let centroid = centroid?;
+    //     Some((centroid.x(), centroid.y()))
+    // }
+
+    /// Option 2 - Get the middle point on the way
+    ///
+    /// The slowest call is the iterator, since we need to know all nodes.
+    /// Which drives a single seek cursor over the nodes table.
+    fn mid_point_on_way(way: &Way, locations: &Locations) -> Option<(f64, f64)> {
+        // Fetch all nodes, driving the iterator to completion at once
+        let positions = way.nodes().collect::<Vec<_>>();
+
+        // Find the mid point, on the line. So the position will be on the line,
+        // which might be off the line somewhere.
+        let mid_position = positions.get(positions.len() / 2)?;
+
+        // Lookup each position
+        let location = locations.get(*mid_position)?;
+
+        Some((location.lat(), location.lon()))
+    }
+
+    // /// Option 3 - Get the first point on the way
```

Comment:
nit: if unused, remove commented code

Reply:
Ah yep, I'd left this in if you had a preference on the approach to use? The are all slowed by the way lookup, which was why I started poking around with the PBF import. Like you suggested previously, I think including a bbox at some stage would be gold.

Reply:
If it doesn't dramatically change total import time let's stick with centroid. There are some optimizations we could make here but it doesn't seem right to me to pick a single node and look up only that one location.

Reply:
Specifically, as far as optimizations go: nodes in a way are often clustered reasonably well in the space of IDs, so if we made a list of (node, S2 cell) pairs and then sorted them by node, that lookup would be pretty cache-friendly.

Reply:
Sorry you reminded me, won't the centroid approach effectively be off the line?

Reply:
Most of the ways that airmail indexes are closed ways representing building outlines so being off of the actual way geometry is fine. How much does this impact total import time? I'm open to whichever approach but if it doesn't impact total import time dramatically then centroid would be my preference. Otherwise we can go with whatever's fast.

Reply:
Although I just realized that this isn't just used for WOF lookups, it's used for the lat/lng returned by the server right? I think it's important that we provide a proper centroid in the search response.

Reply:
It's used for ways, so a user would be centred to that point, but it's also used for the PIP query, right? So for buildings the ux would be fine, but you'd likely be looking far afield for most roads. This might be something to come back to or investigate further, as ideally - a street would be the middle point, but a building the centroid. But to detect which one it is (assuming the last and first points are the same) you have to receive all nodes in both cases and so incur the performance penalty regardless.

I'll set it to centroid for now and update the comments roughly per this?

Reply:
Airmail doesn't currently index roads I think. The highway tag is excluded IIRC. I'm sure there are open ways that we do index but for the most part I think it's just buildings right now. And yeah that sounds great. :)